Make default_data_collator more flexible and deprecate old behavior

sgugger opened the pull request:
This PR does two things:

- avoid breaking changes when people had a custom `DataCollator` with a `collate_batch` method (there is still the breaking change with `DataCollator` not being a class anymore)
- makes `default_data_collator` more flexible by handling dicts on top of `InputExamples` classes coming from our examples

This fixes #5049

sgugger commented on src/transformers/data/data_collator.py:
```diff
@@ -34,30 +34,50 @@ def default_data_collator(features: List[InputDataClass]) -> Dict[str, torch.Ten
     first = features[0]
+    batch = {}
+
+    # Note: below there are a few mysterious torch.stack([torch.tensor(f) for f ...]).
```
This things happens four times below, I put `# See comment in the intro` notes to document it but maybe the corresponding code deserves its own function with the code documented there? Let me know what you prefer.

julien-c commented on src/transformers/data/data_collator.py:
```diff
@@ -34,30 +34,50 @@ def default_data_collator(features: List[InputDataClass]) -> Dict[str, torch.Ten
+    # This is because torch.tensor does not accept a list of tensors, to deal with the case the f already are tensors
+    # and not lists of ints.
+
+    if isinstance(first, dict):
```
This code looks very duplicated but I don't know if we have an easy + performant way of removing the duplication.
Maybe just doing:
```
d = [vars(f) for f in features]
```
at the top and only handle dicts?

sgugger replied:
Oh smart, should indeed work to remove the dupe code.

julien-c replied:
I just hope the perf penalty is acceptable, given that this is on the "hot path" of each batch during training. cc @mfuntowicz

sgugger replied:
Why would it be longer that the current loop over keys of `var(first)` then `getattr`?

sgugger commented:
Refactored on @julien-c suggestion and removed the little hack to handle tensors and lists of ints (which was cauding a 4x slowdown on my tests). I just use the first features to test if we have a tensor (then stack) or not (then use torch.tensor).

julien-c commented:
LGTM!
